contracts/smart-account/modules/SessionValidationModules/ABISessionValidationModule.sol

```diff
@@ -115,46 +127,64 @@ contract ABISessionValidationModule is ISessionValidationModule {
+     * @param permission Permission object that contains rules to be checked
+     * @return true if the calldata matches the permission, false otherwise
+     */
+    function _checkRulesForPermission(
         bytes calldata data,
-        Permission memory permission
+        Permission calldata permission
     ) internal pure returns (bool) {
         if (bytes4(data[0:4]) != permission.selector) return false;
         uint256 length = permission.rules.length;
         for (uint256 i; i < length; ) {
-            Rule memory rule = permission.rules[i];
-            bytes32 param = bytes32(data[4 + rule.offset:4 + rule.offset + 32]);
-            if (rule.condition == Condition.EQUAL && param != rule.value) {
-                return false;
-            } else if (
-                rule.condition == Condition.LESS_THAN_OR_EQUAL &&
-                param > rule.value
-            ) {
-                return false;
-            } else if (
-                rule.condition == Condition.LESS_THAN && param >= rule.value
-            ) {
-                return false;
-            } else if (
-                rule.condition == Condition.GREATER_THAN_OR_EQUAL &&
-                param < rule.value
-            ) {
-                return false;
-            } else if (
-                rule.condition == Condition.GREATER_THAN && param <= rule.value
-            ) {
-                return false;
-            } else if (
-                rule.condition == Condition.NOT_EQUAL && param == rule.value
-            ) {
+            Rule calldata rule = permission.rules[i];
+            Condition condition = rule.condition;
+            uint256 offset = rule.offset;
+            bytes32 value = rule.value;
+            bytes32 param = bytes32(data[4 + offset:4 + offset + 32]);
+
+            bool rulePassed;
+            assembly ("memory-safe") {
+                switch condition
+                case 0 {
+                    // Condition.EQUAL
+                    rulePassed := eq(param, value)
+                }
+                case 1 {
+                    // Condition.LESS_THAN_OR_EQUAL
+                    rulePassed := or(lt(param, value), eq(param, value))
+                }
+                case 2 {
+                    // Condition.LESS_THAN
+                    rulePassed := lt(param, value)
+                }
+                case 3 {
+                    // Condition.GREATER_THAN_OR_EQUAL
+                    rulePassed := or(gt(param, value), eq(param, value))
+                }
+                case 4 {
+                    // Condition.GREATER_THAN
+                    rulePassed := gt(param, value)
+                }
+                case 5 {
+                    // Condition.NOT_EQUAL
+                    rulePassed := not(eq(param, value))
+                }
+            }
+
+            if (!rulePassed) {
                 return false;
             }
+
             unchecked {
                 ++i;
             }
```

Comment:
This optimisation is redundant post 0.8.22
https://soliditylang.org/blog/2023/10/25/solidity-0.8.22-release-announcement/

Feel free to remove it if you like.